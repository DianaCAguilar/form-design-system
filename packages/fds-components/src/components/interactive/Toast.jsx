import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import FDS from 'fds-dictionary/dist/js/styleConstants';
import CheckIcon from 'fds-icons/lib/react/CheckIcon';
import InformationIcon from 'fds-icons/lib/react/InformationIcon';
import DenyIcon from 'fds-icons/lib/react/DenyIcon';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';
import customPropTypes from '../../util/customPropTypes';

const Toast = ({
  content,
  type,
  actionLabel,
  progress,
  onAction,
  // onDismiss,
  // isAutoDismiss,
  canDismiss,
  dismissToast /* Passed from Toaster */,
}) => {
  const classNames = cx('toast', `toast--${type}`);

  let icon;
  switch (type) {
    case 'info':
      icon = <InformationIcon size="l" color={FDS.COLOR_AQUA} />;
      break;
    case 'success':
      icon = <CheckIcon size="xs" color={FDS.COLOR_WHITE} />;
      break;
    case 'warn':
      icon = (
        <div role="image" className="typemod--white">
          !
        </div>
      );
      break;
    case 'error':
      icon = (
        <div role="image" className="typemod--white">
          !
        </div>
      );
      break;
    default:
      icon = null;
  }

  const useContinuousProgressBar = type === 'progress' && progress;
  const useAmountProgressBar = type === 'progress' && !progress;

  const toastBar = (
    <div className="toast-bar">
      {useContinuousProgressBar && (
        <div
          className="toast-progressBar toast-progressBar--amount"
          style={{ width: `${progress}%` }}
        />
      )}
      {useAmountProgressBar && (
        <div className="toast-progressBar toast-progressBar--continuous" />
      )}
    </div>
  );

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      {toastBar}
      <div className="toast-content">
        <Flex align="center">
          {icon && (
            <FlexItem shrink>
              <div className="toast-icon media--s">{icon}</div>
            </FlexItem>
          )}
          <FlexItem>{content}</FlexItem>
          {actionLabel && onAction && (
            <FlexItem shrink>
              <Button className="toast-action" onClick={onAction}>
                {actionLabel}
              </Button>
            </FlexItem>
          )}
          {canDismiss && (
            <FlexItem shrink>
              {/* TODO: pull this out into a separate component with timer */}
              <IconButton onClick={dismissToast}>
                <DenyIcon size="xs" color={FDS.FONT_COLOR_DEFAULT} />
              </IconButton>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

Toast.defaultProps = {
  canDismiss: true,
  isAutoDismiss: true,
};

Toast.propTypes = {
  /** JSX Content of Toast */
  content: PropTypes.element.isRequired,

  /** Type of toast */
  type: PropTypes.oneOf(['info', 'warn', 'error', 'success', 'progress']).isRequired,

  /** Label for action button */
  actionLabel: PropTypes.string,

  /** Callback for action button click */
  onAction: PropTypes.func,

  /** Callback invoked on dismissal of toast */
  onDismiss: PropTypes.func,

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,

  /** Delay auto-dismissal of toast by this delay (in MS) */
  dismissDelay: PropTypes.Number,

  /** Is this toast user-dismissable? */
  canDismiss: PropTypes.bool,

  /** Number 1-100 declaring % progress */
  progress: customPropTypes.range(1, 100),

  /** Passed by Toaster: Number 1-100 declaring % progress of dismissal countdown */
  dismissProgress: customPropTypes.range(1, 100),

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast: PropTypes.func,
};

export default Toast;
