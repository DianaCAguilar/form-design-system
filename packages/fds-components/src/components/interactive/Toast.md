### Info Toast

```js
<Toast
  content={<p className="typemod--light">Check out this <strong className="typemod--dark">cool toast</strong></p>}
  actionLabel="Action"
  onAction={() => {}}
  type="info"
/>
```

### Warn Toast

```js
<Toast
  content={<p>You better <strong className="typemod--dark">watch out</strong></p>}
  actionLabel="run away"
  onAction={() => {}}
  type="warn"
/>
```

### Success Toast

```js
<Toast
  content={<p>You <strong className="typemod--dark">Did it!</strong></p>}
  type="success"
/>
```

### Error Toast

```js
<Toast
  content={<p>You made a <strong className="typemod--dark">huge mistake</strong></p>}
  actionLabel="Try again"
  onAction={() => {}}
  type="error"
/>
```

### Progress Toast (without progress specified)
By default, progress toasts show a continuous animation of a progress bar.
To make the progress bar reflect a specific amount of progress, use the `progress` prop (see below).

```js
<Toast
  content={<p>Uploading a photo of your <strong className="typemod--dark">your cat</strong></p>}
  type="progress"
/>
```

### Progress Toast (with progress specified)

```js
<Toast
  content={<p><strong className="typemod--dark">42%</strong> complete uploading your cat photo</p>}
  progress={42}
  type="progress"
/>
```
