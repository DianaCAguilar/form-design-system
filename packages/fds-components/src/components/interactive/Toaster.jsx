import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

/**
 * @module Toaster
 * Controller for showing, dismissing and positioning Toast components.
 */
const Toaster = ({ delay, toast }) => {
  const [isToasting, setIsToasting] = useState(Boolean(toast));

  /**
   * Removes a toast from `toasts` internal state array
   * @param {String} toastId
   */
  const dismissToast = () => {
    setIsToasting(false);
  };

  // pass dismiss function down to toast
  if (toast) {
    React.cloneElement(toast, { dismissToast });
  }

  // set a timeout of `delay` on `toast` to dismiss itself
  useEffect(() => {
    const timer = setTimeout(dismissToast, delay);

    return function cleanup() {
      clearTimeout(timer);
    };
  });

  return <div className="toaster">{isToasting && toast}</div>;
};

Toaster.defaultProps = {
  delay: 3000,
};

Toaster.propTypes = {
  /** Length of time in MS toasts should auto-dismiss */
  delay: PropTypes.number,

  /** a `Toast` element */
  toast: PropTypes.element,
};

export default Toaster;
