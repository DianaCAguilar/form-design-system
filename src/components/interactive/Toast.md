### Info Toast

```js
<Toast
  content={<p>Check out this <strong className="typemod--dark">toast</strong></p>}
  actionLabel="Action"
  onAction={() => {}}
  type="info"
/>
```

### Warn Toast

```js
<Toast
  content={<p>You better <strong className="typemod--dark">watch out</strong></p>}
  actionLabel="run away"
  onAction={() => {}}
  type="warn"
/>
```

### Success Toast

```js
<Toast
  content={<p>You <strong className="typemod--dark">Did it!</strong></p>}
  type="success"
/>
```

### Error Toast

```js
<Toast
  content={<p>Failed to <strong className="typemod--dark">do a thing</strong></p>}
  actionLabel="Retry"
  onAction={() => {}}
  type="error"
/>
```

### Progress Toast (without progress specified)
By default, progress toasts show a continuous animation of a progress bar.

To make the progress bar reflect a specific amount of progress, use the `progress` prop (see below).

Most of the operations on `cbi-site` are not cancellable, so progress toasts do not render
a dismiss button to avoid misleading the user (dismissing a toast is not the same as
cancelling an operation).

```js
<Toast
  content={<p>Uploading a photo of your <strong className="typemod--dark">your cat</strong></p>}
  type="progress"
/>
```

### Progress Toast (with progress specified)

```js
<Toast
  content={<p><strong className="typemod--dark">42%</strong> complete uploading your cat photo</p>}
  progress={42}
  type="progress"
/>
```
