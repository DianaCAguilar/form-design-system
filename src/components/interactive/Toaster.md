Controller for positioning, transition, and dismissal of toasts.

Timer starts immediately on mount.

**⚠️ Include only one `Toaster` component per page/container** to ensure that only one toast is rendered at a time.