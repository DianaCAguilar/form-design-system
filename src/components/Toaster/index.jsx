import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import customPropTypes from 'components/util/customPropTypes';
import Toast, { TYPES } from './Toast';

const dismissDelay = 4000;

export const Toaster = ({
  isAutoDismiss = true,
  isOpen = false,
  toastInstance = {},
  id,
}) => {
  useEffect(() => {
    // We need to explicitly check of canDismiss is NOT true, as it has a default
    // parameter of true that this component does not know about
    if (
      isAutoDismiss &&
      toastInstance.type !== 'progress' &&
      toastInstance.canDismiss !== false &&
      toastInstance.onDismiss
    ) {
      const timer = setTimeout(() => toastInstance.onDismiss(), dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }

    return undefined;
  }, [
    id,
    isOpen,
    isAutoDismiss,
    toastInstance.type,
    toastInstance.canDismiss,
    toastInstance.onDismiss,
  ]);

  return (
    <div aria-live="assertive">
      <TransitionGroup>
        {isOpen && (
          <CSSTransition
            key={id}
            appear
            unmountOnExit
            timeout={380}
            classNames="rtgSlideIn"
          >
            <div className="toaster">
              <Toast
                dismissDelay={dismissDelay}
                isAutoDismiss={isAutoDismiss}
                dismissToast={toastInstance.onDismiss}
                content={toastInstance.content}
                type={toastInstance.type}
                progress={toastInstance.progress}
                canDismiss={toastInstance.canDismiss}
                actionLabel={toastInstance.actionLabel}
                onAction={toastInstance.onAction}
                dismissOnAction={toastInstance.dismissOnAction}
              />
            </div>
          </CSSTransition>
        )}
      </TransitionGroup>
    </div>
  );
};

const ToasterWrapper = (props) =>
  ReactDOM.createPortal(
    <Toaster {...props} />,
    // eslint-disable-next-line no-undef
    document.body
  );

Toaster.propTypes = {
  /**  A UUID for the toast. Each new toast should have a new UUID. */
  id: PropTypes.string,
  /** Specifies whether Toast is open or not */
  isOpen: PropTypes.bool,
  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,
  /** Controls the display for your specific toast. */
  toastInstance: PropTypes.shape({
    /** JSX Content of Toast */
    content: PropTypes.node.isRequired,

    /** Type of toast */
    type: PropTypes.oneOf(TYPES),

    /** Label for action button */
    actionLabel: PropTypes.string,

    /** Callback for action button click */
    onAction: PropTypes.func,

    /* Controls if toast gets dismiss when the ActionButton is clicked */
    dismissOnAction: PropTypes.bool,

    /** Is this toast user-dismissable? */
    canDismiss: PropTypes.bool,

    /** Number 1-100 declaring % progress */
    progress: customPropTypes.range(1, 100),

    /** Function to dismiss the toast */
    dismissToast: PropTypes.func,
  }),
};

export default ToasterWrapper;
