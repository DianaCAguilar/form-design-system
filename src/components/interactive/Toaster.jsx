import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import uuidv4 from 'uuid/v4';
import Toast from './Toast';

let transitionID = uuidv4();

export const Toaster = ({
  toastProps = {},
  dismissDelay = 4000,
  isAutoDismiss = true,
  isOpen = false,
  onDismiss,
}) => {
  const [toast, setToast] = useState(null);
  const prevContent = useRef();

  const coreLogic = () => {
    if (isOpen) {
      setToast(
        <CSSTransition
          key={transitionID}
          appear
          unmountOnExit
          timeout={380}
          classNames="rtgSlideIn"
        >
          <div className="toaster">
            <Toast
              dismissDelay={dismissDelay}
              dismissToast={onDismiss}
              isAutoDismiss={isAutoDismiss}
              content={toastProps.content}
              type={toastProps.type}
              progress={toastProps.progress}
              canDismiss={toastProps.canDismiss}
              actionLabel={toastProps.actionLabel}
              onAction={toastProps.onAction}
              dismissOnAction={toastProps.dismissOnAction}
            />
          </div>
        </CSSTransition>
      );
    } else {
      setToast(null);
    }

    // We need to explicitly check of canDismiss is NOT true, as it has a default
    // parameter of true that this component does not know about
    if (
      isAutoDismiss &&
      toastProps.type !== 'progress' &&
      toastProps.canDismiss !== false
    ) {
      const timer = setTimeout(() => onDismiss(), dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }

    return undefined;

    // We need to check for all toastProps and all Toaster props here so that
    // an update of them re-renders the Toast (without animation).
    // Note: you "cannot" spread props here... do not attempt to do so :)
    // You may be able to spread "propTypes" from Toast if you're ambitious.
  };

  useEffect(() => {
    console.log('component remounted');
  }, []);

  useEffect(() => {
    transitionID = uuidv4();
    // We check for the props here that we want to re-render the toast animation
  }, [toastProps.content, toastProps.type]);

  useEffect(coreLogic, [
    toastProps.type,
    toastProps.progress,
    toastProps.actionLabel,
    toastProps.canDismiss,
    toastProps.onAction,
    toastProps.dismissOnAction,
    isAutoDismiss,
    dismissDelay,
  ]);

  useEffect(() => {
    // This logic is specifically meant to handle not re-rendering
    // the toast if the same html is passed into content
    const content = JSON.stringify(toastProps.content);
    const didContentChange = content !== prevContent.current;
    if (didContentChange && prevContent.current !== undefined) coreLogic();
    prevContent.current = content;
  }, [toastProps.content]);

  return (
    <div aria-live="assertive">
      <TransitionGroup>{toast}</TransitionGroup>
    </div>
  );
};

const ToasterWrapper = (props) =>
  ReactDOM.createPortal(
    <Toaster {...props} />,
    // eslint-disable-next-line no-undef
    document.body
  );

Toaster.propTypes = {
  /** Specifies whether Toast is open or not */
  isOpen: PropTypes.bool,
  /** Accepts all props that `Toast` accepts. Refer to Toast component for full documentation */
  toastProps: PropTypes.shape(Toast.propTypes),
  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,
  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,
  /** Callback that is fired when toast is dismissed */
  onDismiss: PropTypes.func,
};

export default ToasterWrapper;
