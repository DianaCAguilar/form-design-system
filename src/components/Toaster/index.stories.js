import React, { useState } from 'react';
import { v4 } from 'uuid';

import { Toaster } from 'components/Toaster';
import Toast, { TYPES } from './Toast';

export const Primary = (args) => {
  const [isToastShowing, setIsToastShowing] = useState(false);
  return (
    <>
      <button onClick={() => setIsToastShowing(!isToastShowing)}>
        {isToastShowing ? 'Hide Toaster' : 'Show Toaster'}
      </button>
      <Toaster
        {...args}
        isOpen={isToastShowing}
        onDismiss={() => setIsToastShowing(false)}
        toastProps={{
          content: 'hey',
          type: 'info',
          progress: 42,
          dismissOnAction: true,
          actionLabel: 'Action',
        }}
      />
    </>
  );
};

export const rendering = () => {
  const defaults = { id: v4(), isOpen: false, isAutoDismiss: true };
  const [params, setParams] = useState(defaults);
  const [isShadow, setIsShadow] = useState(false);
  const dismissToast = () => {
    setParams({ ...params, isOpen: false });
  };
  const randomToastType = () => {
    const randomNumber = Math.floor(Math.random() * TYPES.length);
    return TYPES[randomNumber];
  };
  const randomNum = () => Math.floor(Math.random() * 100);

  const { id, type, isOpen, content, isAutoDismiss, actionLabel, progress } = params;
  return (
    <div className={isShadow ? 'elevation--1 padding--all' : 'padding--all'}>
      <button onClick={() => setIsShadow(!isShadow)}>
        {isShadow ? 'Hide Arbitrary Style' : 'Show Arbitrary Style'}
      </button>
      <button onClick={() => setParams({ ...params, isOpen: true })}>Open Toast</button>
      <button onClick={() => setParams({ ...params, isOpen: false })}>Close Toast</button>
      <button onClick={() => setParams({ ...params, isAutoDismiss: !isAutoDismiss })}>
        Toggle isAutoDismiss {isAutoDismiss ? 'off' : 'on'}
      </button>
      <button onClick={() => setParams({ ...params, id: v4() })}>New ID</button>
      <button onClick={() => setParams({ ...params, type: randomToastType() })}>
        Random Toast Type
      </button>
      <button
        onClick={() =>
          setParams({
            ...params,
            actionLabel: actionLabel === 'Hello' ? 'World' : 'Hello',
          })
        }
      >
        Change Action Label
      </button>
      <button onClick={() => setParams({ ...params, content: <b>hey</b> })}>
        Add HTML content
      </button>
      <button onClick={() => setParams({ ...params, actionLabel: null })}>
        Remove Action Label
      </button>
      <button onClick={() => setParams({ ...params, progress: randomNum() })}>
        Add random progress
      </button>
      <br />
      <button
        onClick={() =>
          setParams({
            ...params,
            id: v4(),
            type: 'error',
            isOpen: true,
            content: 'New ID? Yes!',
          })
        }
      >
        Error Toast (new id)
      </button>
      <button
        onClick={() =>
          setParams({
            ...params,
            id: v4(),
            type: 'info',
            isOpen: true,
            content: 'New ID? Yes!',
          })
        }
      >
        Info Toast (new id)
      </button>
      <button
        onClick={() =>
          setParams({
            ...params,
            id,
            type: 'error',
            isOpen: true,
            content: 'New ID? No.',
          })
        }
      >
        Error Toast (no new id)
      </button>
      <button onClick={() => setParams({ ...defaults })}>Reset all</button>
      <br />
      <br />
      <b>NOTE: this is state is EXTERNAL to toaster and is being passed in</b>
      <pre>{JSON.stringify(params, null, 1)}</pre>
      <Toaster
        id={id}
        isOpen={isOpen}
        onDismiss={dismissToast}
        isAutoDismiss={isAutoDismiss}
        toastProps={{
          actionLabel,
          progress,
          type,
          content,
        }}
      />
    </div>
  );
};

export const variations = () => (
  <React.Fragment>
    <Toast
      content={
        <p>
          Check out this <strong className="color--heading">toast</strong>
        </p>
      }
      actionLabel="Action"
      onAction={() => {}}
      type="info"
    />
    <br />
    <Toast
      content={
        <p>
          You better <strong className="color--heading">watch out</strong>
        </p>
      }
      actionLabel="run away"
      onAction={() => {}}
      type="warn"
    />
    <br />
    <Toast
      content={
        <p>
          You <strong className="color--heading">Did it!</strong>
        </p>
      }
      type="success"
    />
    <br />
    <Toast
      content={
        <p>
          Failed to <strong className="color--heading">do a thing</strong>
        </p>
      }
      actionLabel="Retry"
      onAction={() => {}}
      type="error"
    />
    <br />
    <Toast
      content={
        <p>
          Uploading a photo of your <strong className="color--heading">your cat</strong>
        </p>
      }
      type="progress"
    />
    <br />
    <Toast
      content={
        <p>
          <strong className="color--heading">42%</strong> complete uploading your cat
          photo
        </p>
      }
      progress={42}
      type="progress"
    />
  </React.Fragment>
);

export default {
  component: Toaster,
  subcomponents: { Toast },
  title: 'components/Toaster',
  parameters: {
    docs: {
      description: {
        component:
          'Controller for positioning, transition, and dismissal of toasts. Important notes: - **⚠️ Import only one `Toaster` component per application**. - Pass in a brand new `id` (with optional props) if you would like to animate a new toast in. - Passing in new props without changing the id will not retrigger the animation',
      },
    },
  },
};
