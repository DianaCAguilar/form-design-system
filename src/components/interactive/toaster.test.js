import React from 'react';
import { mount, shallow } from 'enzyme';
import uuid from 'uuid/v4';

import Toast from './Toast';
import Toaster from './Toaster';

jest.mock('uuid/v4');

const renderToaster = (useShallow, props) => {
  const toast = <Toast type="info" content={<p>test</p>} />;
  let result;
  if (useShallow) {
    result = shallow(<Toaster toast={toast} {...props} />);
  } else {
    result = mount(<Toaster toast={toast} {...props} />);
  }
  return result;
}


describe('Toaster component', () => {

  uuid.mockImplementation(() => 'testid');

  it('matches snapshot', () => {
    const component = renderToaster(true);
    expect(component).toMatchSnapshot();
  });

  it('passes dismiss func to toast', () => {
    const component = renderToaster();
    const {dismissToast} = component.find(Toast).props();
    expect(typeof dismissToast).toBe('function');
  })

});
