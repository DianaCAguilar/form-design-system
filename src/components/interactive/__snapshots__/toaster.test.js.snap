// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster component matches snapshot 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    aria-live="assertive"
  >
    <TransitionGroup
      childFactory={[Function]}
      component="div"
    >
      <CSSTransition
        appear={true}
        classNames="rtgSlideIn"
        key="testid"
        timeout={380}
        unmountOnExit={true}
      >
        <div
          className="toaster"
        />
      </CSSTransition>
    </TransitionGroup>
  </div>
</Portal>
`;
